pd-client/client.go

```diff
@@ -14,106 +14,320 @@
+		quit:          make(chan struct{}),
+	}
+
+	if err := c.initClusterID(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := c.updateLeader(); err != nil {
 		return nil, errors.Trace(err)
 	}
-	return &client{worker: worker}, nil
+	log.Infof("[pd] init cluster id %v", c.clusterID)
+
+	c.wg.Add(2)
+	go c.tsLoop()
+	go c.leaderLoop()
+
+	// TODO: Update addrs from server continuously by using GetMember.
+
+	return c, nil
 }
 
-func (c *client) Close() {
-	c.worker.stop(errors.New("[pd] pd-client closing"))
+func (c *client) initClusterID() error {
+	for i := 0; i < maxInitClusterRetries; i++ {
+		for _, u := range c.urls {
+			client, err := apiutil.NewClient(u, pdTimeout)
+			if err != nil {
+				log.Errorf("[pd] failed to get cluster id: %v", err)
+				continue
+			}
+			clusterID, err := client.GetClusterID()
+			if err != nil {
+				log.Errorf("[pd] failed to get cluster id: %v", err)
+				continue
+			}
+			c.clusterID = clusterID
+			return nil
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return errors.New("failed to get cluster id")
 }
 
-func (c *client) GetClusterID() uint64 {
-	return c.worker.clusterID
+func (c *client) updateLeader() error {
+	for _, u := range c.urls {
+		client, err := apiutil.NewClient(u, pdTimeout)
+		if err != nil {
+			continue
+		}
+		leader, err := client.GetLeader()
+		if err != nil {
+			continue
+		}
+		c.mu.RLock()
+		changed := c.leader != leader.GetAddr()
+		c.mu.RUnlock()
+		if changed {
+			if err = c.switchLeader(leader.GetAddr()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	}
+	return errors.Errorf("failed to get leader from %v", c.urls)
 }
 
-func (c *client) GetTS() (int64, int64, error) {
-	req := &tsoRequest{
-		done: make(chan error, 1),
+func (c *client) switchLeader(addr string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	log.Infof("[pd] leader switches to: %v, previous: %v", addr, c.leader)
+	if _, ok := c.clients[addr]; !ok {
+		cc, err := grpc.Dial(addr, grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+			u, err := url.Parse(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			return net.Dial(u.Scheme, u.Host)
+		}), grpc.WithInsecure())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.clients[addr] = pdpb2.NewPDClient(cc)
 	}
+	c.leader = addr
+	return nil
+}
+
+func (c *client) leaderLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.checkLeaderCh:
+		case <-time.After(time.Minute):
+		case <-c.quit:
+			return
+		}
 
+		if err := c.updateLeader(); err != nil {
+			log.Errorf("[pd] failed updateLeader: %v", err)
+		}
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case first := <-c.tsoRequests:
+			c.procTSORequests(first)
+		case <-c.quit:
+			return
+		}
+	}
+}
+
+func (c *client) procTSORequests(first *tsoRequest) {
 	start := time.Now()
-	c.worker.requests <- req
-	err := <-req.done
+	ctx, cancel := context.WithTimeout(context.Background(), pdTimeout)
```

Comment:
pass the context outer, the same for leaderLoop 

you can create a cancellable context in NewClient with `context.Background`, save the cancel function, and pass the context to the leaderLoop and tsoLoop. 

In Close function, you can call the cancel function to let the loop return ASAP.